cloud-controller-manager/do/kubernetes_lb_validator.go

```diff
@@ -0,0 +1,176 @@
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/digitalocean/godo"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	lbIDAnnotation string = "kubernetes.digitalocean.com/load-balancer-id"
+)
+
+// LBService represents the v1.service lb object data
+type LBService struct {
+	APIVersion string
+	Kind       string
+	Metadata   Metadata
+	Spec       map[string]interface{}
+	Selector   map[string]interface{}
+	Type       string
+}
+
+// Metadata represents the metadata field in the lb service object
+type Metadata struct {
+	Annotations interface{} `json:"annotation data,omitempty"`
+	Name        string
+	Namespace   string
+}
+
+// DOKSLBServiceValidator validates service type LB
+type KubernetesLBServiceValidator struct {
+	decoder *admission.Decoder
+	Log     logr.Logger
+	GClient *godo.Client
+	Region  string
+}
+
+// DOKSLBServiceValidator ...
+func (v *KubernetesLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+
+	v.Log.V(6).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "failed to decode request")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+	}
+
+	v.Log.V(6).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer || svc.DeletionTimestamp != nil {
+		return admission.Allowed("ignoring the service which is either not a load balancer or is being deleted")
+	}
+
+	// TODO: these forwarding rules are a placeholder. Further development is required to extract the values from the svc object
+	forwardingRules := []godo.ForwardingRule{
+		{
+			EntryProtocol:  "http",
+			EntryPort:      80,
+			TargetProtocol: "http",
+			TargetPort:     80,
+			CertificateID:  "",
+			TlsPassthrough: false,
+		}}
+
+	v.Region, err = dropletRegion(v.GClient.Regions)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to determine region: %v", err))
+	}
+
+	lbRequest := v.buildRequest(svc.Name, v.Region, forwardingRules)
+
+	var resp *godo.Response
+
+	// check if old service object exists
+	if req.OldObject.Raw != nil {
+		// decode raw object
+		err = v.decoder.DecodeRaw(req.OldObject, svc)
+		if err != nil {
+			v.Log.Error(err, "failed to decode existing object")
+			return admission.Errored(http.StatusBadRequest, fmt.Errorf("failed to decode request: %v", err))
+		}
+
+		currentLBID := svc.Annotations[lbIDAnnotation]
+		// perform update if associated with created lb
+		if (len(svc.Status.LoadBalancer.Ingress) > 0 && svc.Status.LoadBalancer.Ingress[0].IP != "") || currentLBID != "" {
+			v.Log.Info(fmt.Sprintf("updating lb id: %v", currentLBID))
+			resp, err = v.validateUpdate(ctx, currentLBID, lbRequest)
+			if resp == nil {
```

Comment:
I don't think we need to check `resp`: godo always returns a non-nil error when the returned HTTP code is >= 400, meaning we should be able to work with `err` only.

Similar for the create path.

Reply:
I thought this would resolve this [comment](https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/586#discussion_r1176092293):

Reply:
Sorry, I meant to move that check into `isValidationError`: that way, the function is safe to be called no matter what `resp` looks like.

Reply:
Does [this](https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/586/files#diff-0b037574b49ca1618f3243d7b959eb12569ec91a928aaa9eab1d10859e31092dR158) refactor address the check?